0000-richer-documents.rst

```diff
@@ -0,0 +1,304 @@
+the "annotation-free" nature of the resulting document.  We could even imagine
+tweaking the functions that actually print documents to only accept
+annotation-free documents, so as to force users to interpret annotations one
+way or another before getting the documents printed somewhere.
+
+In fact, GHC uses ``SDoc`` in other contexts than error messages, a major
+one being code generation. In those cases, we will never want to emit
+any annotation, just pure text, and as fast as possible. This would be another
+case where we would want to deal with ``SDoc Void`` values. It is therefore
+desired that any potential implementation of this proposal doesn't drastically
+change the performance document construction and rendering when no annotation is
+involved.
+
+Pieces of code that don't produce annotations in documents can either
+produce documents that are completely polymorphic in the annotation type,
+or ``SDoc Void`` values. The polymorphic variant has the merit of being
+usable in code that produces documents with a specific annotation type. This
+would let us print snippets of expressions inside error messages
+(``SDoc ErrAnnotation``), or in the output generated by some ``-ddump-*`` flag
+(``SDoc Void``).
+
+However, with the changes described so far, we would run into a problem. While
+the semantics of the existing ``Doc`` constructors are pretty clear in terms of
+layout, there is no good answer when interpreting a ``Pure a``. We will never
+know for sure what text this annotation is going to end up being replaced with,
+since the point of this proposal exactly consists in sprinkling non-textual
+Haskell values all over our documents and rendering them later. This in turns
+means that we can't reliably "guess" whether our annotation is going to end up
+being rendered over several lines, nor how many columns or levels of nesting it
+will involve.
+
+To work around this problem, we suggest to adopt the trick used in the
+`wl-pprint-extras <https://hackage.haskell.org/package/wl-pprint-extras>`_
+library, which consists in introducing constructors that allow users to
+introduce (sub-)documents that are dependent on the exact column number,
+nesting level, ribbon length, etc. For instance, this is how we would define
+the constructor that "suspends" a sub-document on the column number of the
+current position in the textual rendering of a larger document:::
+
+    data Doc a
+      = ...
+      | Pure a
+      | Column (Int -> Doc a) -- new!
+      | ...
+
+While this now prevents ``Doc`` from being ``Foldable`` or ``Traversable``
+(which isn't that big of a deal since ``SDoc`` wasn't going to support those
+operations anyway and it is the type that we manipulate the most),
+it nicely solves the problem of having to "guess" properties about the
+textual rendition of an annotation, while mixing several pieces of documents
+to produce a larger one using specific semantics.
```

Comment:
I still want more information about e.g. `collect`-like operations. How will clients access the annotations? Is it worth thinking/talking about returning annotations with source locations?

Reply:
I'm not done with that section. :-)

I'm waiting for the corresponding "sub-discussion" to come to a concrete conclusion that I can then expand into a proper explanation as well as examples.